cuda/context_manager.hpp

```diff
@@ -0,0 +1,135 @@
+ *   2. cuLibraryGetModule
+ *   3. cuCtxPopCurrent_v2_(entry|exit)
+ *
+ * Cleanup:
+ *   1. cuDevicePrimaryCtxRelease_(entry|exit)
+ */
+class CUDAContextManager {
+public:
+  // returns true and sets output var if found, false otherwise
+  bool get_device(hpt_t hpt, CUdevice *pdev);
+
+  // returns true and sets output var if found, false otherwise
+  bool get_top_context(hpt_t hpt, CUcontext *pctx);
+
+  // Note: streams associate with a context (and a device) at create time; at time
+  // of use, they may be using a device other than the one associated with the
+  // thread's top context stack.
+  // returns true and sets output var if found, false otherwise.
+  // if stream is 0 / default, fallback to get_device (top context device)
+  bool get_stream_device(hpt_t hpt, CUstream stream, CUdevice *pdev);
+
+  // save dev
+  void primary_ctx_retain_entry(hpt_t hpt, CUdevice dev);
+  void primary_ctx_retain_exit(hpt_t hpt, CUresult cuResult, CUcontext ctx);
+
+  void primary_ctx_release_entry(hpt_t hpt, CUdevice dev);
+  void primary_ctx_release_exit(hpt_t hpt, CUresult cuResult);
+
+  void primary_ctx_reset_entry(hpt_t hpt, CUdevice dev);
+  void primary_ctx_reset_exit(hpt_t hpt, CUresult cuResult);
+
+  void ctx_create_entry(hpt_t hpt, CUdevice dev);
+  void ctx_create_exit(hpt_t hpt, CUresult cuResult, CUcontext ctx);
+
+  void ctx_destroy_entry(hpt_t hpt, CUcontext ctx);
+  void ctx_destroy_exit(hpt_t hpt, CUresult cuResult);
+
+  void ctx_set_current_entry(hpt_t hpt, CUcontext ctx);
+  void ctx_set_current_exit(hpt_t hpt, CUresult cuResult);
+
+  void ctx_push_current_entry(hpt_t hpt, CUcontext ctx);
+  void ctx_push_current_exit(hpt_t hpt, CUresult cuResult);
+
+  void ctx_pop_current_entry(hpt_t hpt);
+  void ctx_pop_current_exit(hpt_t hpt, CUresult cuResult, CUcontext ctx);
+
+  // Note: entry callback not needed
+  void stream_create_exit(hpt_t hpt, CUresult cuResult, CUstream phStream_val);
+
+private:
+  void ensure_stack(hpt_t hpt);
+
+  // There is at most once primary context per (host, process, device)
+  std::unordered_map<hp_device_t, CUcontext> hp_device_primary_ctx_;
+
+  // Contexts must be on the per thread stack to be used, even primary contexts
+  std::unordered_map<hpt_t, context_stack_t> hpt_ctx_stack_;
+
+  // Contexts can be used by any thread in a process, regardless of which
+  // thread they were created in, so they are not dependent on thread id
+  std::unordered_map<hp_context_t, CUdevice> hp_ctx_to_device_;
+
+  // Track streams so when they are passed to API calls we know what context
+  // (and thereby what device) the API call is using
+  std::unordered_map<hp_stream_t, CUcontext> hp_stream_to_ctx_;
+
+  // Temp saves to carry args from entry to exit (since we need to wait for exit
+  // callback and check cuResult before doing anything). Need to keep track per thread.
+  std::unordered_map<hpt_t, CUcontext> hpt_entry_ctx_;
+  std::unordered_map<hpt_t, CUdevice> hpt_entry_dev_;
```

Comment:
Should we use an `entry_state` for those one?

Reply:
Not sure what is going on with github, but latest version uses entry_state and does not have these.